src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs

```diff
@@ -201,15 +192,31 @@ internal static bool TryUseHttps(this ListenOptions listenOptions)
     /// <returns>The <see cref="ListenOptions"/>.</returns>
     public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConnectionAdapterOptions httpsOptions)
     {
-        var loggerFactory = listenOptions.KestrelServerOptions?.ApplicationServices.GetRequiredService<ILoggerFactory>() ?? NullLoggerFactory.Instance;
+        return listenOptions.UseHttps(new Lazy<HttpsConnectionAdapterOptions>(httpsOptions));
+    }
 
+    /// <summary>
+    /// Configure Kestrel to use HTTPS. This does not use default certificates or other defaults specified via config or
+    /// <see cref="KestrelServerOptions.ConfigureHttpsDefaults(Action{HttpsConnectionAdapterOptions})"/>.
+    /// </summary>
+    /// <param name="listenOptions">The <see cref="ListenOptions"/> to configure.</param>
+    /// <param name="lazyHttpsOptions">Options to configure HTTPS.</param>
+    /// <returns>The <see cref="ListenOptions"/>.</returns>
+    private static ListenOptions UseHttps(this ListenOptions listenOptions, Lazy<HttpsConnectionAdapterOptions> lazyHttpsOptions)
+    {
         listenOptions.IsTls = true;
-        listenOptions.HttpsOptions = httpsOptions;
+        listenOptions.HttpsOptions = lazyHttpsOptions;
 
+        // NB: This lambda will only be invoked if either HTTP/1.* or HTTP/2 is being used
         listenOptions.Use(next =>
         {
+            // Evaluate the HttpsConnectionAdapterOptions, now that the configuration, if any, has been loaded
+            var httpsOptions = listenOptions.HttpsOptions.Value;
             // Set the list of protocols from listen options
             httpsOptions.HttpProtocols = listenOptions.Protocols;
```

Comment:
Doing this here would be an issue if the HTTP/3 logic needed to reference `HttpsOptions.HttpProtocols`. We could just pass the HttpProtocols directly to the middleware constructor and remove `HttpsOptions.HttpProtocols` to make it clear that this isn't needed for the HTTP/3 path. make it clear that this isn't needed. We could also just move this line into the lazy callback.

Reply:
It is very hard to keep track of the source of truth for each of these properties and the order in which they need to be accessed.

Reply:
To confirm, you're saying this was a problem with the code before my change, right?

Reply:
Now passing the protocols to the middleware directly.